From f540b805f6dd0bedf059a6e232e904f8966200e9 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:06:12 -0400
Subject: [virt] kvm/ppc: book3s hv - Fix vcore leak

Message-id: <1410545655-205645-144-git-send-email-dzickus@redhat.com>
Patchwork-id: 94155
O-Subject: [RHEL7 PATCH 143/626] kvm: powerpc: book3s hv: Fix vcore leak
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit f1378b1c0bdce349e21f271dec0612a6cfac5d14
Author: Paul Mackerras <paulus@samba.org>
Date:   Fri Sep 27 15:33:43 2013 +0530

    kvm: powerpc: book3s hv: Fix vcore leak

    add kvmppc_free_vcores() to free the kvmppc_vcore structures
    that we allocate for a guest, which are currently being leaked.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index 59cf70f..5857c1b 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -2050,10 +2050,20 @@ int kvmppc_core_init_vm(struct kvm *kvm)
  return 0;
 }
 
+static void kvmppc_free_vcores(struct kvm *kvm)
+{
+ long int i;
+
+ for (i = 0; i < KVM_MAX_VCORES; ++i)
+  kfree(kvm->arch.vcores[i]);
+ kvm->arch.online_vcores = 0;
+}
+
 void kvmppc_core_destroy_vm(struct kvm *kvm)
 {
  uninhibit_secondary_onlining();
 
+ kvmppc_free_vcores(kvm);
  if (kvm->arch.rma) {
   kvm_release_rma(kvm->arch.rma);
   kvm->arch.rma = NULL;
-- 
1.7.1